support-frontend/assets/pages/contributions-landing/contributionsLandingInit.ts

```diff
@@ -167,32 +164,47 @@ function selectInitialAmounts(
 ) {
 	const { amounts } = state.common;
 	const amountFromUrl = getAmountFromUrl();
-	Object.keys(amounts).forEach((contributionType) => {
+
+	const amountForContributionType = (contributionType: ContributionType) => {
 		if (amountFromUrl && contributionType === selectedContributionType) {
 			if (
 				amountFromUrl !== 'other' &&
 				amounts[contributionType].amounts.includes(amountFromUrl)
 			) {
-				dispatch(selectAmount(amountFromUrl, contributionType));
-			} else {
-				dispatch(selectAmount('other', contributionType));
-
-				if (amountFromUrl !== 'other') {
-					dispatch(updateOtherAmount(`${amountFromUrl}`, contributionType));
-				}
+				return amountFromUrl;
 			}
-		} else {
-			const { defaultAmount } = amounts[contributionType];
-			dispatch(selectAmount(defaultAmount, contributionType));
+			return 'other';
 		}
-	});
+		return amounts[contributionType].defaultAmount;
+	};
+
+	dispatch(
+		selectAmounts({
+			ONE_OFF: amountForContributionType('ONE_OFF'),
+			MONTHLY: amountForContributionType('MONTHLY'),
+			ANNUAL: amountForContributionType('ANNUAL'),
+		}),
+	);
+
+	// This means there is a query parameter specifying an amount,
+	// but that amount isn't available as one of the choice cards.
+	// The logic in 'amountForContributionType' will select 'other'
+	// in this case. Additionally, we also want to prefill the 'other'
+	// field with the amount in the url.
+	if (
+		amountFromUrl &&
+		amountFromUrl !== 'other' &&
+		!amounts[selectedContributionType].amounts.includes(amountFromUrl)
+	) {
+		dispatch(updateOtherAmount(`${amountFromUrl}`, selectedContributionType));
+	}
```

Comment:
By factoring out the `amountForContributionType` function, we now have to have this separate check + dispatch, but I think it makes it a bit more explicit as to what is going on.

Reply:
fair enough and I agree it makes the dispatches clearer, my only concern here is we have duplicated between line 197 and 172 now, so we need to make sure they keep in step in future.
Not a major thing, but I wonder if there's a way of returning a boolean when the check is done the first time, and once we know it's "other" then we don't need to redo the check?